coordinator/persistence-cassandra-4.1/src/main/java/io/stargate/db/cassandra/impl/idempotency/IdempotencyAnalyzer.java

```diff
@@ -0,0 +1,138 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.stargate.db.cassandra.impl.idempotency;
+
+import com.datastax.oss.driver.shaded.guava.common.collect.ImmutableSet;
+import java.util.Set;
+import org.apache.cassandra.cql3.CQLStatement;
+import org.apache.cassandra.cql3.Lists;
+import org.apache.cassandra.cql3.Operation;
+import org.apache.cassandra.cql3.functions.Function;
+import org.apache.cassandra.cql3.functions.TimeFcts;
+import org.apache.cassandra.cql3.functions.UuidFcts;
+import org.apache.cassandra.cql3.statements.BatchStatement;
+import org.apache.cassandra.cql3.statements.DeleteStatement;
+import org.apache.cassandra.cql3.statements.ModificationStatement;
+import org.apache.cassandra.cql3.statements.SelectStatement;
+import org.apache.cassandra.cql3.statements.TruncateStatement;
+import org.apache.cassandra.cql3.statements.UseStatement;
+import org.apache.cassandra.db.marshal.CounterColumnType;
+import org.apache.cassandra.db.marshal.ListType;
+import org.apache.cassandra.schema.ColumnMetadata;
+import org.apache.cassandra.schema.SchemaTransformation;
+
+/**
+ * The IdempotencyAnalyzer responsibility is to analyze `CqlStatement`s and infer whether the query
+ * is idempotent or not. All prepared queries are analyzed. If yes, it is safe to retry such a query
+ * on the client or/and server-side. If it is not, the query should not be retried because it may
+ * result in an inconsistent state of a Database.
+ *
+ * <ul>
+ *   The analyzer returns that query is non-idempotent in the following scenarios: List appends,
+ *   prepends, or element deletions.
+ *   <li>Counter increments / decrements.
+ *   <li>Mutations that use non-idempotent functions such as `now()` and `uuid()`
+ *   <li>LWTs
+ *   <li>Truncate, schema changes, and USE
+ * </ul>
+ *
+ * <p>Batches are idempotent if their underlying statements are all idempotent. It is done by
+ * examining queries within a `BatchStatement`. It is worth noting that all reads are idempotent.
+ */
+public class IdempotencyAnalyzer {
+  private static final Set<Function> NON_IDEMPOTENT_FUNCTION;
+
+  static {
+    NON_IDEMPOTENT_FUNCTION =
+        ImmutableSet.<Function>builder().addAll(TimeFcts.all()).addAll(UuidFcts.all()).build();
+  }
+
+  public static boolean isIdempotent(CQLStatement statement) {
+    // if any of the BatchStatement is non-idempotent, return false
+    if (statement instanceof BatchStatement) {
+      BatchStatement batchStatement = (BatchStatement) statement;
+      for (ModificationStatement s : batchStatement.getStatements()) {
+        boolean isIdempotent = analyzeStatement(s);
+        if (!isIdempotent) {
+          return false;
+        }
+      }
+    }
```

Comment:
I think this is/was copied from 4.0 pretty much as-is...  but it seems to be this is missing a

    return true;

after iterating over all statements within batch? If all are deemed idempotent, there is a separate call to "analyzeStatement(...)". Which, I think, does actually return `true` in the end after going past all `if`/`else` statements.

So it guess it works, just does a bit of unnecessary work?